### PR TITLE
Create Meetup Page

### PR DESCRIPTION
**Pull request by jamieliu386**
Fixes: #25

Connects the frontend to the backend.

Had to move the CORS middleware from `setupMiddlewares` to `setupGlobalMiddleware` because when creating a meetup, the client first sends an `OPTIONS` request to the server to retrieve the http methods available for the `/meetup` endpoint. Originally, the most external router only allows for `GET` and `POST` to hit that endpoint, so it returned an error. By moving the CORS middleware to `setupGlobalMiddleware`, we circumvent this error by hitting the CORS middleware before the request reaches the router.

TODO:
- redirect to the meetup page once it is created

### CODE REVIEW

File: client/src/components/CreateMeetup/CreateMeetup.js [hunk 1]
```diff
@@ -33,15 +35,47 @@ const useStyles = makeStyles(() => ({
...
+
+  const clearForm = () => {
+    setTitle("");
+    setTime(new Date());
+    setMeetupType("");
+    setMeetupURL("");
+    setMeetupLocation(null);
+    setGroupCount([2, 10]);
+    setDescription("");
+    setTags([]);
```
Comment (TimothyGu) [inline client/src/components/CreateMeetup/CreateMeetup.js:58]: Do we also want to reset error and creatingMeetups?

File: client/src/components/CreateMeetup/CreateMeetup.js [hunk 2]
```diff
@@ -98,11 +149,11 @@ function CreateMeetup() {
             value={meetupType}
             onChange={(event) => setMeetupType(event.target.value)}
           >
-            <MenuItem value="in-person">In person</MenuItem>
-            <MenuItem value="remote">Remote</MenuItem>
+            <MenuItem value={IN_PERSON}>In person</MenuItem>
+            <MenuItem value={REMOTE}>Remote</MenuItem>
```
Comment (TimothyGu) [inline client/src/components/CreateMeetup/CreateMeetup.js:153]: +1

File: client/src/lib/fetch.js [hunk 3]
```diff
@@ -9,6 +9,55 @@ export async function getUserData(id) {
   return { res, resJSON: await res.json() };
 }
 
-export const doOtherThing = () => {
+export function blah() {
```
Comment (TimothyGu) [inline client/src/lib/fetch.js:12]: Delete

File: client/src/lib/fetch.js [hunk 4]
```diff
@@ -9,6 +9,55 @@ export async function getUserData(id) {
...
+    location = {
+      url: meetupURL,
+    };
+  }
+
+  const meetup = {
+    description,
+    location,
+    maxCapacity: groupCount[1],
+    minCapacity: groupCount[0],
```
Comment (TimothyGu) [inline client/src/lib/fetch.js:47]:
```suggestion
    minCapacity: groupCount[0],
    maxCapacity: groupCount[1],
```

File: client/src/lib/fetch.js [hunk 5]
```diff
@@ -9,6 +9,55 @@ export async function getUserData(id) {
...
+    tags,
+    time: time.toISOString(),
```
Comment (TimothyGu) [inline client/src/lib/fetch.js:49]:
```suggestion
    time,
```

See:

```js
JSON.stringify(new Date()) === `"2020-11-30T04:37:41.869Z"`
```

File: client/src/lib/fetch.js [hunk 6]
```diff
@@ -9,6 +9,55 @@ export async function getUserData(id) {
...
+  meetupLocation,
+  groupCount,
+  description,
+  tags,
+}) {
+  const createMeetupEndpoint = `${SERVER_URL}/meetup`;
+
+  let location;
+  if (meetupType === IN_PERSON) {
+    location = {
```
Comment (TimothyGu) [inline client/src/lib/fetch.js:30]: Can you also include the name of the location? It should be available as a part of meetupLocation.

File: server/impl/meetup.go [hunk 7]
```diff
@@ -150,7 +150,7 @@ func (i *Implementation) GetMeetupID(params operations.GetMeetupIDParams) middle
 	if id := SessionFromContext(ctx).Values[UserID]; id == nil {
```
Comment (TimothyGu) [inline server/impl/meetup.go:150]: If you could do me a favor and rename `idStr` to `userID`, that'd be very appreciated.

File: server/restapi/configure_down_to_meet.go [hunk 8]
```diff
@@ -6,11 +6,12 @@ import (
 	"crypto/tls"
 	"net/http"
 
+	"github.com/rs/cors"
+
 	"github.com/go-openapi/runtime"
 	"github.com/go-openapi/runtime/middleware"
 	"github.com/go-openapi/runtime/security"
 	"github.com/go-openapi/swag"
```
Comment (TimothyGu) [inline server/restapi/configure_down_to_meet.go:14]:
```suggestion
	"github.com/go-openapi/runtime"
	"github.com/go-openapi/runtime/middleware"
	"github.com/go-openapi/runtime/security"
	"github.com/go-openapi/swag"
	"github.com/rs/cors"
```

File: server/restapi/configure_down_to_meet.go [hunk 9]
```diff
@@ -111,9 +118,3 @@ func setupMiddlewares(handler http.Handler) http.Handler {
 	c.Log = log.WithField("source", "cors")
 	return impl.RequestMiddleware(Impl.SessionMiddleware(c.Handler(handler)))
```
Comment (TimothyGu) [inline server/restapi/configure_down_to_meet.go:119]: Can you just move the CORS handler and impl.RequestMiddleware here, but leave Impl.SessionMiddleware in setupMiddlewares?

The CORS handler has to be in setupGlobalMiddleware, that's fine. impl.RequestMiddleware is for logging, so it's fine too. SessionMiddleware is only used for the routes, so should stay in the original function.

File: client/src/components/CreateMeetup/CreateMeetup.js [hunk 10]
```diff
@@ -210,11 +254,25 @@ function CreateMeetup() {
     );
   };
 
+  const onSubmit = () => {
+    setCreatingMeetup(true);
+    if (!validateForm()) {
+      setCreatingMeetup(false);
+      return;
+    }
+    createMeetup();
```
Comment (TimothyGu) [inline client/src/components/CreateMeetup/CreateMeetup.js:263]:
```suggestion
    if (creatingMeetup || !validateForm()) {
      return;
    }
    setCreatingMeetup(true);
    createMeetup();
```

Also, feel free to just inline the createMeetup function here. No need for it to be a separate function.

**Comment by TimothyGu** [comment]
Could you explain a bit in the PR description why you had to move the CORS middleware to setupGlobalMiddleware?

File: client/src/lib/fetch.js (hunk 3 above)
Comment (jamieliu386) [reply client/src/lib/fetch.js:12]: oops :3

File: server/impl/meetup.go (hunk 7 above)
Comment (jamieliu386) [reply server/impl/meetup.go:150]: `userID` is already being used in some places though

**Review by TimothyGu — APPROVED** [review]

**Pull request merged**